queue/queue.mbt

```diff
@@ -0,0 +1,333 @@
+// Copyright 2024 International Digital Economy Academy
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+struct Cont[T] {
+  content : T
+  mut next : Cell[T]
+}
+
+enum Cell[T] {
+  Nil
+  Cons(Cont[T])
+}
+
+struct Queue[T] {
+  mut length : Int
+  mut first : Cell[T]
+  mut last : Cell[T]
+}
+
+/// Creates a new empty queue.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::new()
+/// ```
+pub fn Queue::new[T]() -> Queue[T] {
+  { length: 0, first: Nil, last: Nil }
+}
+
+test "new" {
+  let queue : Queue[Int] = new()
+  @assertion.assert_eq(queue.length, 0)?
+}
+
+/// Creates a new queue from an array.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::from_array([1, 2, 3, 4])
+/// ```
+pub fn Queue::from_array[T](arr : Array[T]) -> Queue[T] {
+  if arr.length() == 0 {
+    return new()
+  }
+  let queue : Queue[T] = Queue::new()
+  queue.length = arr.length()
+  for i = arr.length(); i > 0; i = i - 1 {
+    let cell : Cell[T] = Cons({ content: arr[i - 1], next: queue.first })
+    queue.first = cell
+  }
+  queue.last = queue.first
+  queue
+}
+
+/// Tests if two queue cells are equal.
+pub fn op_equal[T : Eq](self : Cell[T], other : Cell[T]) -> Bool {
+  loop self, other {
+    Nil, Nil => true
+    Cons({ content: x, next: xs }), Cons({ content: y, next: ys }) => {
+      if x != y {
+        return false
```

Comment:
loop is break by default, the return is not needed, removing return will make the control flow more clear, easier for compiler optimizations later